File: src/main/java/pl/training/workshop/m8/s02_stranglerfig/step4/BookingModule.java
package pl.training.workshop.m8.s02_stranglerfig.step4;

import pl.training.workshop.m8.s02_stranglerfig.BookingLedger;
import pl.training.workshop.shared.Money;

/** Krok 4 (bez zmian): nowy moduł rezerwacji - Money, nazwane reguły, wspólna baza z legacy. */
public final class BookingModule {
    private static final Money ONLINE_FEE = Money.of("2.00");
    private static final int GROUP_SIZE = 10;

    private final BookingLedger ledger;

    public BookingModule(BookingLedger ledger) {
        this.ledger = ledger;
    }

    public String book(String email, String title, int format, int tickets, boolean web) {
        if (tickets <= 0) {
            return "ERROR: no seats";
        }
        Money value = basePrice(format).times(tickets);
        if (tickets >= GROUP_SIZE) {
            value = value.minus(value.percent(10));
        }
        Money fees = web ? ONLINE_FEE.times(tickets) : Money.ZERO;
        String id = ledger.nextId();
        ledger.add(new BookingLedger.Booking(id, email, title, tickets, value, fees));
        return id;
    }

    private static Money basePrice(int format) {
        return switch (format) {
            case 1 -> Money.of("25.00");
            case 2 -> Money.of("32.00");
            default -> Money.of("40.00");
        };
    }
}

File: csharp/src/Training.Workshop/M8/S02StranglerFig/Step4/BookingModule.cs
using Training.Workshop.Shared;

namespace Training.Workshop.M8.S02StranglerFig.Step4;

/// <summary>Krok 4 (bez zmian): nowy moduł rezerwacji - Money, nazwane reguły, wspólna baza z legacy.</summary>
public sealed class BookingModule
{
    private static readonly Money OnlineFee = Money.Of("2.00");
    private const int GroupSize = 10;

    private readonly BookingLedger _ledger;

    public BookingModule(BookingLedger ledger)
    {
        _ledger = ledger;
    }

    public string Book(string email, string title, int format, int tickets, bool web)
    {
        if (tickets <= 0)
        {
            return "ERROR: no seats";
        }
        Money value = BasePrice(format).Times(tickets);
        if (tickets >= GroupSize)
        {
            value = value.Minus(value.Percent(10));
        }
        Money fees = web ? OnlineFee.Times(tickets) : Money.Zero;
        string id = _ledger.NextId();
        _ledger.Add(new BookingLedger.Booking(id, email, title, tickets, value, fees));
        return id;
    }

    private static Money BasePrice(int format) => format switch
    {
        1 => Money.Of("25.00"),
        2 => Money.Of("32.00"),
        _ => Money.Of("40.00"),
    };
}
